visualization/app/codeCharta/ui/codeMap/threeViewer/threeSceneService.ts

```diff
@@ -160,6 +164,124 @@ export class ThreeSceneService implements CodeMapPreRenderServiceSubscriber {
+
+			const norm = Math.sqrt(Math.pow(this.rayPoint.x, 2) + Math.pow(this.rayPoint.y, 2) + Math.pow(this.rayPoint.z, 2))
+			const cameraPoint = raycaster.ray.origin
+			const maxDistance = this.calculateMaxDistance(hoveredLabel, labels, cameraPoint, norm)
+
+			this.normedTransformVector = new Vector3(this.rayPoint.x / norm, this.rayPoint.y / norm, this.rayPoint.z / norm)
+			this.normedTransformVector.multiplyScalar(maxDistance)
+
+			hoveredLabel.position.add(this.normedTransformVector)
+
+			this.modifiedLabel = hoveredLabel
+		}
+	}
+
+	resetLabel() {
+		if (this.modifiedLabel !== null) {
+			this.modifiedLabel.position.sub(this.normedTransformVector)
+			this.modifiedLabel["material"].opacity = this.mapLabelColors.alpha
+			this.modifiedLabel = null
+		}
+	}
+	getLabelForHoveredNode(hoveredBuilding: CodeMapBuilding, labels: Object3D[]) {
+		if (labels === null) {
+			return null
+		}
+		for (let counter = 0; counter < labels.length; counter += 2) {
+			if (labels[counter].userData.node === hoveredBuilding.node) {
+				return labels[counter]
+			}
+		}
+		return null
+	}
+
+	private isOverlapping1D(minBox1: number, maxBox1: number, minBox2: number, maxBox2: number) {
+		return maxBox1 >= minBox2 && maxBox2 >= minBox1
+	}
+
+	private getIntersectionDistance(bboxHoveredLabel: Box3, bboxObstructingLabel: Box3, normedVector: Vector3, distance: number) {
+		normedVector.multiplyScalar(distance)
+		bboxHoveredLabel.translate(normedVector)
+
+		if (
+			(this.isOverlapping1D(bboxObstructingLabel.min.x, bboxObstructingLabel.max.x, bboxHoveredLabel.min.x, bboxHoveredLabel.max.x) &&
+				this.isOverlapping1D(
+					bboxObstructingLabel.min.y,
+					bboxObstructingLabel.max.y,
+					bboxHoveredLabel.min.y,
+					bboxHoveredLabel.max.y
+				)) ||
+			(this.isOverlapping1D(bboxObstructingLabel.min.x, bboxObstructingLabel.max.x, bboxHoveredLabel.min.x, bboxHoveredLabel.max.x) &&
+				this.isOverlapping1D(
+					bboxObstructingLabel.min.z,
+					bboxObstructingLabel.max.z,
+					bboxHoveredLabel.min.z,
+					bboxHoveredLabel.max.z
+				)) ||
+			(this.isOverlapping1D(bboxObstructingLabel.min.y, bboxObstructingLabel.max.y, bboxHoveredLabel.min.y, bboxHoveredLabel.max.y) &&
+				this.isOverlapping1D(
+					bboxObstructingLabel.min.z,
+					bboxObstructingLabel.max.z,
+					bboxHoveredLabel.min.z,
+					bboxHoveredLabel.max.z
+				))
+		) {
+			return distance
+		}
+		return 0
+	}
+
+	private calculateMaxDistance(hoveredLabel: Object3D, labels: Object3D[], cameraPoint: Vector3, norm: number) {
+		let maxDistance = 0
+		for (let counter = 0; counter < labels.length; counter += 2) {
+			const bboxHoveredLabel = new Box3().setFromObject(hoveredLabel)
+			const centerPoint = new Vector3()
+			bboxHoveredLabel.getCenter(centerPoint)
+			const distanceLabelCenterToCamera = cameraPoint.distanceTo(centerPoint)
+			let maxDistanceForLabel = distanceLabelCenterToCamera / 20 //creates a nice small highlighting for hovered, unobstructed labels, empirically gathered value
```

Comment:
Why do you do the calculation for the hoveredLabel multiple times within the loop? Can we do this once instead?